refactor: move several replica_2_9 functions into ar_replica_2_9

JamesPiechota opened the pull request.

JamesPiechota committed:
refactor: move several replica_2_9 functions into ar_replica_2_9

JamesPiechota committed:
refactor: replace ENTROPY_SUB_CHUNK_COUNT with ENTROPY_SIZE

JamesPiechota committed:
refactor: replace get_entropy_bucket_start with get_chunk_bucket_start

JamesPiechota committed:
refactor: drop the "mask" term, I don't think it's needed anymore?

JamesPiechota committed:
docs: write ar_replica_2_9 module docs

also add some tests and rename sub_chunk_index to slice_index

ldmberman commented on apps/arweave/src/ar_chunk_storage.erl:
```diff
@@ -333,16 +348,17 @@ handle_cast(do_prepare_replica_2_9, State) ->
 			false ->
 				false
 		end,
+	%% XXX: SubChunkStart2 will always be equal to SubChunkStart, right?
```
I kept it from the sub-chunk based version where the increments were of the sub-chunk size. Also, I kept the two-element cursor format such that it will be slightly easier to re-introduce sub-chunk based iterations if we want it (e.g., might be helpful to support low-memory setups.)

JamesPiechota replied:
Got it! I'll update the comment.

ldmberman commented on apps/arweave/src/ar_replica_2_9.erl:
```diff
@@ -0,0 +1,473 @@
+    %% Ensuring we have a multiple of 32 entropies proves very convenient for chunk-by-chunk
+    %% syncing.
+    %% 
+    %% Some entropies will be slightly underused.
```
This note is outdated. Every bit of the generated entropy is used now.

JamesPiechota replied:
Updated! https://github.com/ArweaveTeam/arweave/pull/665/commits/eeff94bd8ea1c2d02d219610dee41970ab2d3f54

ldmberman commented on apps/arweave/src/ar_replica_2_9.erl:
```diff
@@ -0,0 +1,473 @@
+    
+    %% In tests the strict data split threshold is 262144 * 3, before that offset chunks
+    %% were not padded. So each provided end offset is taken as is. After the threshold each
+    %% offset is padded to a multiple of ?DATA_CHUNK_SIZE (i.e. 262144).
```
After the threshold is padded to threshold + a multiple of ?DATA_CHUNK_SIZE, to be precise, before determining the entropy bucket (also involves padding.)

JamesPiechota replied:
Updated! https://github.com/ArweaveTeam/arweave/pull/665/commits/eeff94bd8ea1c2d02d219610dee41970ab2d3f54

JamesPiechota committed:
refactor: revert to having a distinct get_entropy_bucket_start/1

ar_chunk_storage:get_chunk_bucket_start/1 is mostly about how we
manage chunk storage and so isn't part of the protocol.

get_entropy_bucket_start, however, is part of the protocol so
best to keep them separate

JamesPiechota committed:
docs: update some comments
